src/interpreter.rs

```diff
@@ -1,20 +1,53 @@
-use crate::ast::{BinaryOperator, Expression, LogicalExpression, Predicate, Value};
+use crate::ast::{BinaryOperator, Predicate, Value};
 use crate::context::{Context, Match};
+use crate::lir::{LirInstruction, LirLogicalOperators, LirProgram};
 
 pub trait Execute {
     fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool;
 }
 
-impl Execute for Expression {
+#[derive(Debug, Clone, Copy)]
+pub enum OperandItem<'a> {
+    Val(bool),
+    Predicate(&'a Predicate),
+}
+
+#[inline]
+fn evaluate_operand_item(item: OperandItem, ctx: &mut Context, m: &mut Match) -> bool {
+    match item {
+        OperandItem::Val(b) => b,
+        OperandItem::Predicate(p) => p.execute(ctx, m),
+    }
+}
+
+impl Execute for LirProgram {
     fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool {
-        match self {
-            Expression::Logical(l) => match l.as_ref() {
-                LogicalExpression::And(l, r) => l.execute(ctx, m) && r.execute(ctx, m),
-                LogicalExpression::Or(l, r) => l.execute(ctx, m) || r.execute(ctx, m),
-                LogicalExpression::Not(r) => !r.execute(ctx, m),
-            },
-            Expression::Predicate(p) => p.execute(ctx, m),
+        let mut operand_stack: Vec<OperandItem> = Vec::new();
```

Comment:
IMO this does not perform better than recursion implementation, plus it cannot short-circuit the right branch of an `and` or `or`.

Reply:
fixed.